src/Utilities/Compiler/Extensions/OperationBlockAnalysisContextExtension.cs

```diff
@@ -60,6 +60,18 @@ static bool IsSingleStatementBody(IBlockOperation body)
                         innerOperation = exprStatement.Operation;
                     }
 
+                    // Special case for methods with a return type and using arrowed expression where the throw-statement
```

Comment:
I am wondering if some ticket shall be created on roslyn's side or if this is the expected behavior. This is not the behavior I was expecting but I can understand why it is like this. 

Note that as the helper is updated we might have some impact on `MarkMembersAsStatic` (the only other rule relying on this helper).

Reply:
Can we instead define a variant of `GetTopmostExplicitDescendants`, say `bool TryGetSingleTopmostExplicitDescendant(IOperation operation, out var singleTopmostExplicitOperation)`: https://github.com/dotnet/roslyn-analyzers/blob/8dd4be4fa590db4512a1089c4df1b970e0d4a527/src/Utilities/Compiler/Extensions/IOperationExtensions.cs#L154-L159

This entire if-body can then probably be implemented as:
```csharp
                    if (TryGetSingleTopmostExplicitDescendant(out var innerOperation) &&
                        innerOperation is IThrowOperation throwOperation &&
                        throwOperation.GetThrownExceptionType() is ITypeSymbol createdExceptionType)
                    {
                        if (Equals(context.Compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemNotImplementedException), createdExceptionType.OriginalDefinition)
                            || Equals(context.Compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemNotSupportedException), createdExceptionType.OriginalDefinition))
                        {
                            return true;
                        }
                    }
```

Reply:
Also see https://github.com/dotnet/roslyn/pull/42408 for similar bug fix in Roslyn.